File: faunadb-scala/src/main/scala/faunadb/FaunaClient.scala
package faunadb

import com.codahale.metrics.MetricRegistry
import com.fasterxml.jackson.databind.{DeserializationFeature, JsonNode, ObjectMapper}
import com.fasterxml.jackson.databind.node.{ArrayNode, NullNode}
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import com.faunadb.common.Connection
import com.faunadb.common.Connection.JvmDriver
import faunadb.errors._
import faunadb.query.{Expr, Get}
import faunadb.values.{ArrayV, Metrics, MetricsResponse, NullV, Value}

import java.io.IOException
import java.net.ConnectException
import java.net.http.HttpResponse
import java.util.concurrent.{CompletionException, Flow, TimeoutException}
import com.faunadb.common.http.ResponseBodyStringProcessor
import faunadb.FaunaClient.EventField
import faunadb.streaming.{BodyValueFlowProcessor, SnapshotEventFlowProcessor}

import scala.collection.JavaConverters._
import scala.compat.java8.DurationConverters._
import scala.compat.java8.FutureConverters._
import scala.compat.java8.OptionConverters._
import scala.concurrent.duration.FiniteDuration
import scala.concurrent.{ExecutionContext, Future}

/** Companion object to the FaunaClient class. */
object FaunaClient {

  /**
    * Creates a new FaunaDB client.
    *
    * @param secret The secret material of the auth key used. See [[https://fauna.com/documentation#authentication-key_access]]
    * @param endpoint URL of the FaunaDB service to connect to. Defaults to https://db.fauna.com
    * @param metrics An optional [[com.codahale.metrics.MetricRegistry]] to record stats.
    * @param queryTimeout An optional global timeout for all the queries issued by this client. The timeout value has
    *                     milliseconds precision. If not provided, a default timeout value is set on the server side.
    * @param userAgent A value used for the User-Agent HTTP header.
    * @return A configured FaunaClient instance.
    */
  def apply(
    secret: String = null,
    endpoint: String = null,
    metrics: MetricRegistry = null,
    queryTimeout: FiniteDuration = null,
    userAgent: String = null,
    checkNewVersion: Boolean = true,
    customHeaders: Map[String, String] = Map.empty): FaunaClient = {

    val b = Connection.builder
    if (endpoint ne null) b.withFaunaRoot(endpoint)
    if (secret ne null) b.withAuthToken(secret)
    if (metrics ne null) b.withMetrics(metrics)
    if (queryTimeout ne null) b.withQueryTimeout(queryTimeout.toJava)
    if (customHeaders.nonEmpty) b.withCustomHeaders(customHeaders.asJava)
    b.withJvmDriver(JvmDriver.SCALA)
    b.withScalaVersion(util.Properties.versionNumberString)
    b.withUserAgent(userAgent)
    b.withCheckNewDriverVersion(checkNewVersion)

    new FaunaClient(b.build)
  }

  sealed abstract class EventField(val value: String)
  case object DocumentField extends EventField("document")
  case object PrevField extends EventField("prev")
  case object DiffField extends EventField("diff")
  case object ActionField extends EventField("action")
}

/**
  * The Scala native client for FaunaDB.
  *
  * Create a new client using [[faunadb.FaunaClient.apply]].
  *
  * Query requests are made asynchronously: All methods will return a
  * [[scala.concurrent.Future]].
  *
  * Example:
  * {{{
  * case class User(ref: RefV, name: String, age: Int)
  *
  * val client = FaunaClient(secret = "myKeySecret")
  *
  * val fut = client.query(Get(Ref(Class("users"), "123")))
  * val instance = Await.result(fut, 5.seconds)
  *
  * val userCast =
  *   for {
  *     ref <- instance("ref").to[RefV]
  *     name <- instance("data", "name").to[String]
  *     age <- instance("data", "age").to[Int]
  *   } yield {
  *     User(ref, name, age)
  *   }
  *
  * userCast.get
  * }}}
  *
  * @constructor create a new client with a configured [[com.faunadb.common.Connection]].
  */
class FaunaClient private (connection: Connection) {

  private[this] val json = new ObjectMapper
  json.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
  json.registerModule(new DefaultScalaModule)

  /**
    * Issues a query.
    *
    * @param expr the query to run, created using the query dsl helpers in [[faunadb.query]].
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @return A [[scala.concurrent.Future]] containing the query result.
    *         The result is an instance of [[faunadb.values.Result]],
    *         which can be cast to a typed value using the
    *         [[faunadb.values.Field]] API. If the query fails, failed
    *         future is returned.
    */
  def query(expr: Expr)(implicit ec: ExecutionContext): Future[Value] = query(expr, None)

  /**
    * Issues a query.
    *
    * @param expr the query to run, created using the query dsl helpers in [[faunadb.query]].
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @param timeout the timeout for the current query. It replaces the timeout value set for this
    *                [[faunadb.FaunaClient]] if any for the scope of this query. The timeout value has
    *                milliseconds precision.
    * @return A [[scala.concurrent.Future]] containing the query result.
    *         The result is an instance of [[faunadb.values.Result]],
    *         which can be cast to a typed value using the
    *         [[faunadb.values.Field]] API. If the query fails, failed
    *         future is returned.
    */
  def query(expr: Expr, timeout: FiniteDuration)(implicit ec: ExecutionContext): Future[Value] = query(expr, Some(timeout))

  /**
    * Issues a query.
    *
    * @param expr the query to run, created using the query dsl helpers in [[faunadb.query]].
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @param timeout the timeout for the current query. It replaces the timeout value set for this
    *                [[faunadb.FaunaClient]] if any for the scope of this query. The timeout value has
    *                milliseconds precision.
    * @return A [[scala.concurrent.Future]] containing the query result.
    *         The result is an instance of [[faunadb.values.Result]],
    *         which can be cast to a typed value using the
    *         [[faunadb.values.Field]] API. If the query fails, failed
    *         future is returned.
    */
  def query(expr: Expr, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[Value] =
    performRequest(json.valueToTree(expr), timeout)

  /**
    * Issues a query.
    *
    * @param expr the query to run, created using the query dsl helpers in [[faunadb.query]].
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @param timeout the timeout for the current query. It replaces the timeout value set for this
    *                [[faunadb.FaunaClient]] if any for the scope of this query. The timeout value has
    *                milliseconds precision.
    * @return A [[scala.concurrent.Future]] containing the query result.
    *         The result is an instance of [[faunadb.values.Result]],
    *         which can be cast to a typed value using the
    *         [[faunadb.values.Field]] API. If the query fails, failed
    *         future is returned.
    */
  def queryWithMetrics(expr: Expr, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[MetricsResponse] =
    performRequestWithMetrics(json.valueToTree(expr), timeout)

  /**
    * Issues multiple queries as a single transaction.
    *
    * @param exprs the queries to run.
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @return A [[scala.concurrent.Future]] containing an IndexedSeq of
    *         the results of each query. Each result is an instance of
    *         [[faunadb.values.Value]], which can be cast to a typed
    *         value using the [[faunadb.values.Field]] API. If *any*
    *         query fails, a failed future is returned.
    */
  def query(exprs: Iterable[Expr])(implicit ec: ExecutionContext): Future[IndexedSeq[Value]] =
    query(exprs, None)

  /**
    * Issues multiple queries as a single transaction.
    *
    * @param exprs the queries to run.
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @param timeout the timeout for the current query. It replaces the timeout value set for this
    *                [[faunadb.FaunaClient]] if any, for the scope of this query. The timeout value
    *                has milliseconds precision.
    * @return A [[scala.concurrent.Future]] containing an IndexedSeq of
    *         the results of each query. Each result is an instance of
    *         [[faunadb.values.Value]], which can be cast to a typed
    *         value using the [[faunadb.values.Field]] API. If *any*
    *         query fails, a failed future is returned.
    */
  def query(exprs: Iterable[Expr], timeout: FiniteDuration)(implicit ec: ExecutionContext): Future[IndexedSeq[Value]] =
    query(exprs, Some(timeout))

  /**
    * Issues multiple queries as a single transaction.
    *
    * @param exprs the queries to run.
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @param timeout the timeout for the current query. It replaces the timeout value set for this
    *                [[faunadb.FaunaClient]] if any, for the scope of this query. The timeout value
    *                has milliseconds precision.
    * @return A [[scala.concurrent.Future]] containing an IndexedSeq of
    *         the results of each query. Each result is an instance of
    *         [[faunadb.values.Value]], which can be cast to a typed
    *         value using the [[faunadb.values.Field]] API. If *any*
    *         query fails, a failed future is returned.
    */
  def query(exprs: Iterable[Expr], timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[IndexedSeq[Value]] =
    performRequest(json.valueToTree(exprs), timeout).map { result =>
      result.asInstanceOf[ArrayV].elems
    }

  private def performRequestCommon[T](body: JsonNode, timeout: Option[FiniteDuration], handler: HttpResponse[String] => Future[T])(implicit ec: ExecutionContext): Future[T] = {
    val javaTimeout = timeout.map(_.toJava).asJava
    val response: Future[HttpResponse[String]] = connection.post("", body, javaTimeout).toScala

    response
      .flatMap {
        case successResponse if successResponse.statusCode() < 300 => handler(successResponse)
        case errorResponse => handleErrorResponse(errorResponse.statusCode(), errorResponse.body())
      }
      .recoverWith(handleNetworkExceptions)
  }

  private def performRequest(body: JsonNode, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[Value] = {
    performRequestCommon(body, timeout, handleSuccessResponse)
  }

  private def performRequestWithMetrics(body: JsonNode, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[MetricsResponse] = {
    performRequestCommon(body, timeout, handleSuccessResponseWithMetrics)
  }

  private def handleSuccessResponseWithMetrics(response: HttpResponse[String])(implicit ec: ExecutionContext): Future[MetricsResponse] = {
    val metricsMap = Metrics.All
      .iterator
      .map(item => item -> response.headers().firstValue(item.toString).asScala)
      .collect {
        case (key, Some(value)) => key -> value
      }
      .toMap

    handleSuccessResponse(response).map(item => MetricsResponse(item, metricsMap))
  }

  /**
    * Creates a subscription to the result of the given read-only expression. When
    * executed, the expression must only perform reads and produce a single
    * streamable type, such as a reference or a version. Expressions that attempt
    * to perform writes or produce non-streamable types will result in an error.
    * Otherwise, any expression can be used to initiate a stream, including
    * user-defined function calls.
    *
    * @param expr the query to subscribe to.
    * @param fields fields to opt-in on the events.
    * @param snapshot if true the second event will be a snapshot event of the target
    * @param ec the `ExecutionContext` used to run the query asynchronously.
    * @return A [[scala.concurrent.Future]] containing a [[java.util.concurrent.Flow.Publisher]] which yields element of
    *         type [[faunadb.values.Value]]. The [[scala.concurrent.Future]] fails if the stream cannot be setup.
    */
  def stream(expr: Expr, fields: Seq[EventField] = Nil, snapshot: Boolean = false)(implicit ec: ExecutionContext): Future[Flow.Publisher[Value]] =
    performStreamRequest(json.valueToTree(expr), fields).map { valuePublisher =>
      if (snapshot) {
        val documentValueFlowProcessor = new SnapshotEventFlowProcessor(() => query(Get(expr)))
        valuePublisher.subscribe(documentValueFlowProcessor)
        documentValueFlowProcessor
      } else {
        valuePublisher
      }
    }

  private def performStreamRequest(body: JsonNode, fields: Seq[EventField] = Nil)(implicit ec: ExecutionContext): Future[Flow.Publisher[Value]] = {
    val params = Map("fields" -> fields.iterator.map(_.value).toList.asJava).asJava
    connection.performStreamRequest("POST", "stream", body, params)
      .toScala
      .flatMap {
        case successResponse if successResponse.statusCode() < 300 =>
          // Subscribe a new FlowEventValueProcessor to consume the Body's Flow.Publisher
          val flowEventValueProcessor = new BodyValueFlowProcessor(json, txn => syncLastTxnTime(txn), () => connection.performStreamRequest("POST", "stream", body, params))
          successResponse.body().subscribe(flowEventValueProcessor)
          Future.successful(flowEventValueProcessor)
        case errorResponse =>
          // The request failed, we need to consume the body manually for error reporting
          ResponseBodyStringProcessor.consumeBody(errorResponse)
            .toScala
            .flatMap(errorBody => handleErrorResponse(errorResponse.statusCode(), errorBody))
      }
      .recoverWith(handleNetworkExceptions)
  }


  /**
    * Creates a new scope to execute session queries. Queries submitted within the session scope will be
    * authenticated with the secret provided. A session client shares its parent's [[com.faunadb.common.Connection]]
    * instance.
    *
    * @param secret user secret for the session scope
    * @param session a function that receives a session client
    * @return the value produced by the session function
    */
  def sessionWith[A](secret: String)(session: FaunaClient => A): A = {
    val client = sessionClient(secret)
    session(client)
  }

  /**
    * Create a new session client. The returned session client shares its parent [[com.faunadb.common.Connection]]
    * instance.
    *
    * @param secret user secret for the session client
    * @return a new session client
    */
  def sessionClient(secret: String): FaunaClient = new FaunaClient(connection.newSessionConnection(secret))

  /**
   * Get the freshest timestamp reported to this client.
   */
  def lastTxnTime: Long = connection.getLastTxnTime

  /**
   * Sync the freshest timestamp seen by this client.
   *
   * This has no effect if staler than currently stored timestamp.
   * WARNING: This should be used only when coordinating timestamps across
   *          multiple clients. Moving the timestamp arbitrarily forward into
   *          the future will cause transactions to stall.
   */
  def syncLastTxnTime(timestamp: Long): Unit =
    connection.syncLastTxnTime(timestamp)

  private def parseResponseBody(responseBody: String)(implicit ec: ExecutionContext): Future[JsonNode] = {
    def parse: Future[Option[JsonNode]] = Future(Option(json.readTree(responseBody)))

    parse.flatMap {
      case Some(json) => Future.successful(json)
      case None => Future.failed(new IOException("Invalid JSON."))
    }
  }

  private def handleSuccessResponse(response: HttpResponse[String])(implicit ec: ExecutionContext): Future[Value] = {
    def getResource(body: JsonNode): Future[JsonNode] = Option(body.get("resource")) match {
      case Some(resource) => Future.successful(resource)
      case None => Future.failed(new IOException("Invalid JSON."))
    }

    def parseValue(resource: JsonNode): Future[Value] = resource match {
      case _: NullNode => Future.successful(NullV)
      case _: JsonNode => Future(json.treeToValue[Value](resource, classOf[Value]))
    }

    for {
      body <- parseResponseBody(response.body())
      resource <- getResource(body)
      value <- parseValue(resource)
    } yield value
  }

  private def handleErrorResponse(statusCode: Int, responseBody: String)(implicit ec: ExecutionContext): Future[Nothing] = {
    def parseErrors(): Future[QueryErrorResponse] = {
      def getErrors(body: JsonNode): Future[Iterator[JsonNode]] = Option(body.get("errors")) match {
        case Some(errors: ArrayNode) => Future.successful(errors.iterator().asScala)
        case _ => Future.successful(Iterator.empty)
      }

      def parseErrors(errors: Iterator[JsonNode]): Future[IndexedSeq[QueryError]] = Future {
        errors.map(json.treeToValue(_, classOf[QueryError])).toIndexedSeq
      }

      val result: Future[QueryErrorResponse] =
        for {
          body <- parseResponseBody(responseBody)
          errors <- getErrors(body)
          queryErrors <- parseErrors(errors)
        } yield QueryErrorResponse(statusCode, queryErrors)

      result
        .recoverWith {
          case e: FaunaException => Future.failed(e)
          case unavailable if statusCode == 503 => Future.failed(new UnavailableException("Service Unavailable: Unparseable response.", unavailable))
          case unknown => Future.failed(new UnknownException(s"Unparseable service $unknown response.", unknown))
        }
    }

    def parseErrorsAndFailWith(fun: QueryErrorResponse => FaunaException): Future[Nothing] = {
      parseErrors().flatMap { errors =>
        val exception = fun(errors)
        Future.failed(exception)
      }
    }

    statusCode match {
      case 400 => parseErrorsAndFailWith(new BadRequestException(_))
      case 401 => parseErrorsAndFailWith(new UnauthorizedException(_))
      case 403 => parseErrorsAndFailWith(new PermissionDeniedException(_))
      case 404 => parseErrorsAndFailWith(new NotFoundException(_))
      case 409 => parseErrorsAndFailWith(new TransactionContentionException(_))
      case 500 => parseErrorsAndFailWith(new InternalException(_))
      case 503 => parseErrorsAndFailWith(new UnavailableException(_))
      case _   => parseErrorsAndFailWith(new UnknownException(_))
    }
  }

  private def handleNetworkExceptions[A]: PartialFunction[Throwable, Future[A]] = {
    case ex: ConnectException =>
      Future.failed(new UnavailableException(ex.getMessage, ex))
    case ex: TimeoutException =>
      Future.failed(new UnavailableException(ex.getMessage, ex))
    case ex: CompletionException if ex.getCause.isInstanceOf[IOException] && ex.getMessage.contains("header parser received no bytes") =>
      Future.failed(new UnavailableException(ex.getMessage, ex))
    case ex: CompletionException if ex.getCause.isInstanceOf[IOException] && ex.getMessage.contains("too many concurrent streams") =>
      Future.failed(BadRequestException(None, "the maximum number of streams has been reached for this client"))
  }

}

File: faunadb-android/src/main/java/com/faunadb/client/FaunaClient.java
package com.faunadb.client;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.faunadb.client.errors.*;
import com.faunadb.client.query.Expr;
import com.faunadb.client.types.Field;
import com.faunadb.client.types.Value;

import java.io.IOError;
import java.io.IOException;
import java.net.ConnectException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiConsumer;
import java.util.function.Function;

import okhttp3.*;

import static com.faunadb.client.types.Codec.VALUE;

/**
 * The Java native client for FaunaDB.
 * <p>
 * This client is asynchronous, so all methods that perform latent operations return a {@link CompletableFuture}.
 * <p>
 * Queries are constructed by using the static helpers in the {@link com.faunadb.client.query.Language} package.
 * <p>
 * <b>Example</b>:
 * <pre>{@code
 * import static com.faunadb.client.query.Language.*;
 * FaunaClient client = FaunaClient.builder().withSecret("someAuthToken").build();
 * client.query(Get(Ref(Class("some_class"), "ref")));
 * }
 * </pre>
 *
 * @see com.faunadb.client.query.Language
 */
public class FaunaClient {

  private static final int INITIAL_REF_COUNT = 0;
  private static final int DEFAULT_CONNECTION_TIMEOUT_MS = 10000;
  private static final int DEFAULT_REQUEST_TIMEOUT_MS = 60000;
  private static final URL FAUNA_ROOT;

  static {
    try {
      FAUNA_ROOT = new URL("https://db.fauna.com");
    } catch (MalformedURLException e) {
      throw new IOError(e); // won't happen
    }
  }

  private final AtomicInteger refCount;

  /**
   * Creates a new {@link Builder}
   *
   * @return a new {@link Builder}
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * A builder for creating an instance of {@link FaunaClient}
   */
  public static final class Builder {

    private String secret;
    private URL endpoint;
    private OkHttpClient httpClient;

    private Builder() {
    }

    /**
     * Sets the secret to be passed to FaunaDB as a authentication token
     *
     * @param secret the auth token secret
     * @return this {@link Builder} object
     */
    public Builder withSecret(String secret) {
      this.secret = secret;
      return this;
    }

    /**
     * Sets the FaunaDB endpoint url for the built {@link FaunaClient}.
     *
     * @param endpoint the root endpoint URL
     * @return this {@link Builder} object
     * @throws java.net.MalformedURLException when an invalid string is passed
     */
    public Builder withEndpoint(String endpoint) throws MalformedURLException {
      this.endpoint = new URL(endpoint);
      return this;
    }

    /**
     * Sets a custom {@link OkHttpClient} implementation that the built {@link FaunaClient} will use.
     * This custom implementation can be provided to control the behavior of the underlying HTTP transport.
     *
     * @param httpClient the custom {@link OkHttpClient} instance
     * @return this {@link Builder} object
     */
    public Builder withHttpClient(OkHttpClient httpClient) {
      this.httpClient = httpClient;
      return this;
    }

    /**
     * Returns a newly constructed {@link FaunaClient} with configuration based on the settings of this {@link Builder}.
     *
     * @return a new {@link FaunaClient}
     */
    public FaunaClient build() {
      URL faunaEndpoint = FAUNA_ROOT;
      if (endpoint != null)
        faunaEndpoint = endpoint;

      if (httpClient == null) {
        return new FaunaClient(new OkHttpClient.Builder()
                .connectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS, TimeUnit.MILLISECONDS)
                .readTimeout(DEFAULT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
                .writeTimeout(DEFAULT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
                .retryOnConnectionFailure(true)
                .build(), faunaEndpoint, secret);
      }

      return new FaunaClient(httpClient, faunaEndpoint, secret);
    }
  }

  private final ObjectMapper json = new ObjectMapper().findAndRegisterModules();
  private final OkHttpClient client;
  private final URL endpoint;
  private final String authHeader;

  private FaunaClient(OkHttpClient client, URL endpoint, String secret) {
    this(new AtomicInteger(INITIAL_REF_COUNT), client, endpoint, secret);
  }

  private FaunaClient(AtomicInteger refCount, OkHttpClient client, URL endpoint, String secret) {
    this.refCount = refCount;
    this.client = client;
    this.endpoint = endpoint;
    this.authHeader = Credentials.basic(secret, "");

    this.refCount.incrementAndGet();
  }

  /**
   * Creates a session client with the user secret provided. Queries submited to a session client will be
   * authenticated with the secret provided.
   *
   * @param secret user secret for the session client
   * @return a new {@link FaunaClient}
   */
  public FaunaClient newSessionClient(String secret) {
    if (refCount.get() > 0)
      return new FaunaClient(refCount, client, endpoint, secret);
    else
      throw new IllegalStateException("Can not create a session connection from a closed http connection");
  }

  /**
   * Frees any resources held by the client. Also closes the underlying {@link OkHttpClient}.
   *
   */
  public void close() {
    if (refCount.decrementAndGet() == 0) {
      client.dispatcher().executorService().shutdown();
      client.connectionPool().evictAll();
    }
  }

  /**
   * Issues a Query to FaunaDB.
   * <p>
   * Queries are modeled through the FaunaDB query language, represented by the helper functions in the
   * {@link com.faunadb.client.query.Language} class.
   * <p>
   * Responses are modeled as a general response tree. Each node is a {@link Value}, and
   * can be converted to structured types through various methods on that class.
   *
   * @param expr The query expression to be sent to FaunaDB.
   * @return A {@link CompletableFuture} containing the root node of the Response tree.
   * @see Value
   * @see com.faunadb.client.query.Language
   */
  public CompletableFuture<Value> query(Expr expr) {
    return performRequest(json.valueToTree(expr));
  }


  /**
   * Issues multiple queries to FaunaDB.
   * <p>
   * These queries are sent to FaunaDB in a single request. A list containing all responses is returned
   * in the same order as the issued queries.
   * <p>
   *
   * @param exprs the list of queries to be sent to FaunaDB.
   * @return a {@link CompletableFuture} containing an ordered list of the query's responses.
   */
  public CompletableFuture<List<Value>> query(Expr... exprs) {
    return query(Arrays.asList(exprs));
  }

  /**
   * Issues multiple queries to FaunaDB.
   * <p>
   * These queries are sent to FaunaDB in a single request, and are evaluated. The list of response nodes is returned
   * in the same order as the issued queries.
   * <p>
   * See {@link FaunaClient#query(Expr)} for more information on the individual queries.
   *
   * @param exprs the list of query expressions to be sent to FaunaDB.
   * @return a {@link CompletableFuture} containing an ordered list of root response nodes.
   */
  public CompletableFuture<List<Value>> query(List<? extends Expr> exprs) {
      return performRequest(json.valueToTree(exprs)).thenApply(new Function<Value, List<Value>>() {
              public List<Value> apply(Value result) {
                  return result.collect(Field.as(VALUE));
              }
          });
  }

  private CompletableFuture<Value> performRequest(JsonNode body) {
    try {
      Request request = new Request.Builder()
              .url(endpoint)
              .post(RequestBody.create(MediaType.parse("application/json"), json.writeValueAsString(body)))
              .addHeader("Authorization", authHeader)
              .addHeader("X-FaunaDB-API-Version", "2.1")
              .build();

      final CompletableFuture<Value> rv = new CompletableFuture();

      client.newCall(request).enqueue(new Callback() {
        @Override
        public void onFailure(Call call, IOException ex) {
          rv.completeExceptionally(ex);
        }

        @Override
        public void onResponse(Call call, Response response) throws IOException {
          try {
            handleQueryErrors(response);

            JsonNode responseBody = parseResponseBody(response);
            JsonNode resource = responseBody.get("resource");
            rv.complete(json.treeToValue(resource, Value.class));
          } catch (Exception ex) {
            rv.completeExceptionally(ex);
          }
        }
      });

      return handleNetworkExceptions(rv);
    } catch (Exception ex) {
        CompletableFuture<Value> oops = new CompletableFuture();
        oops.completeExceptionally(ex);
        return oops;
    }
  }

  private void handleQueryErrors(Response response) throws FaunaException {
    int status = response.code();
    if (status >= 300) {
      try {
        ArrayNode errors = (ArrayNode) parseResponseBody(response).get("errors");
        List<HttpResponses.QueryError> errorBuilder = new ArrayList<>();

        for (JsonNode errorNode : errors) {
          errorBuilder.add(json.treeToValue(errorNode, HttpResponses.QueryError.class));
        }

        HttpResponses.QueryErrorResponse errorResponse = HttpResponses.QueryErrorResponse.create(status, errorBuilder);

        switch (status) {
          case 400:
            throw new BadRequestException(errorResponse);
          case 401:
            throw new UnauthorizedException(errorResponse);
          case 403:
            throw new PermissionDeniedException(errorResponse);
          case 404:
            throw new NotFoundException(errorResponse);
          case 500:
            throw new InternalException(errorResponse);
          case 503:
            throw new UnavailableException(errorResponse);
          default:
            throw new UnknownException(errorResponse);
        }
      } catch (IOException ex) {
        switch (status) {
          case 503:
            throw new UnavailableException("Service Unavailable: Unparseable response.");
          default:
            throw new UnknownException("Unparseable service " + status + "response.");
        }
      }
    }
  }

  private <V> CompletableFuture<V> handleNetworkExceptions(CompletableFuture<V> f) {
    return f.whenComplete(new BiConsumer<V, Throwable>() {
      public void accept(V v, Throwable ex) {
        if (ex == null) {
          return;
        } else if (ex instanceof ConnectException ||
                   ex instanceof TimeoutException) {
          throw new UnavailableException(ex.getMessage());
        }
      }
    });
  }

  private JsonNode parseResponseBody(Response response) throws IOException {
    return json.readTree(response.body().string());
  }
}
